### ISSUE TITLE
[Bug]: Category "Amavenita (ship, 2014)" not found

### ISSUE THREAD

**Issue by Ainali** [commons-app/apps-android-commons#4901]
### Summary

I serached for a category that exists, but it doesn't show up.

### Steps to reproduce

1. When uploading an image, search for category: Amavenita

### Expected behaviour

[Category:Amavenita (ship, 2014)](https://commons.wikimedia.org/wiki/Category:Amavenita_(ship,_2014)) should be suggested.

### Actual behaviour

"No Categories found"

### Device name

Samsung Galaxy S10

### Android version

Android 12

### Commons app version

3.1.1~1c9267ca0

### Screen-shots

![Screenshot_20220318-180530_Commons](https://user-images.githubusercontent.com/5566159/159050916-23c318f1-f0bb-4c53-b8b9-01a1b24a1e4f.jpg)

**Comment by devarsh-mavani-19** [comment]
@Ainali isn't this the expected behaviour.? Because, Amavenita (ship, 2014) is a non-hidden category and I think we are only supposed to show the hidden category.

**Comment by Ainali** [comment]
Oh, I didn't realize. Then this is a duplicate with #4192 (and a use case for why it is needed, this is how all ships are categorized and the way for a human to identify them).

**Comment by nicolas-raoul** [comment]
> "we are only supposed to show the hidden category in search bar"

@devarsh-mavani-19 I would consider that a bug. If a category is not hidden, then the upload wizard's category search step is **supposed to** show it.

**Comment by nicolas-raoul** [comment]
By the way, when I search for "rabbits", the first result is https://commons.wikimedia.org/wiki/Category:Rabbits which is not a hidden category:

![adb](https://user-images.githubusercontent.com/99590/159115296-dff3836c-346d-4088-b66f-00e2c4f8d59e.png)

**Comment by devarsh-mavani-19** [comment]
> By the way, when I search for "rabbits", the first result is https://commons.wikimedia.org/wiki/Category:Rabbits which is not a hidden category:
> 
> ![adb](https://user-images.githubusercontent.com/99590/159115296-dff3836c-346d-4088-b66f-00e2c4f8d59e.png)

Yes that is how it should work if it is non-hidden, then it should be displayed. However, in case of Ainali it was not displayed because category name contained a year which was older than previous year  (Amavenita (ship, **2014**)). so according to this issue #750 it should be filtered out.

so just to summarize currently 2 conditions must met 1. it should be non-hidden and 2nd either should not have any year in name or the year should be current or previous year.

**Comment by Ainali** [comment]
> However, in case of Ainali it was not displayed because category name contained a year which was older than previous year (Amavenita (ship, **2014**)). so according to this issue #750 it should be filtered out.

Then I would still consider that issue to have been too poorly defined, because the app shouldn't filter out things where the year is part of the identifier of that thing, in this case, the ship that got the name in 2014. (Also, the app should probably not filter out by year, since it's common to upload files a day or two after they are taken which means that if I upload an image on January 1, I expect to see the category for the year before. Safer would be to filter categories where the year is less than current year minus 2.)

**Comment by sivaraam** [comment]
I'm presuming the issue #750 was understood properly but the fix for it was a bit loose than it should've been. The [relevant code snippet](https://github.com/commons-app/apps-android-commons/blob/15fa93e8d004bd0f55515e6d33c7228b23d63026/app/src/main/java/fr/free/nrw/commons/category/CategoriesModel.kt):

```kotlin
return item.matches(".*(19|20)\\d{2}.*".toRegex())
                && !item.contains(yearInString)
                && !item.contains(prevYearInString)
                || item.matches("(.*)needing(.*)".toRegex())
                || item.matches("(.*)taken on(.*)".toRegex())
                || item.matches(".*0s.*".toRegex())
                && !item.matches(".*(200|201)0s.*".toRegex())
```

That essential returns true (which means the category is filtered out) when the name:

1. Contains "19xx" OR "20xx" AND
2. Does not contain current year AND
3. Does not contain previous year OR
4. Contains "needing" OR
5. Contains "take on" OR
6. Contains 0s AND
7. Does not contain "2000s" OR "2010s"

If we were to convert the implicit precedence to explicit ones the condition looks like this:

(((1) AND (2) AND (3)) OR (4) OR (5) OR ((6) AND (7)))

Given [short-circuiting evaluation](https://en.wikipedia.org/wiki/Short-circuit_evaluation), the condition should've returned `true` for "Amavenita (ship, 2014)" thus the category getting filtered out.

We could fix this by fixing the loose condition. I'll open a PR.

**Comment by nicolas-raoul** [comment]
Thanks a lot @devarsh-mavani-19 for finding this out!
I had forgotten about this feature.
Do wee all agree on this?
- `Photographs taken on 2015-11-08` -> Hide
- `Amavenita (ship, 2014)` -> Do not hide

**Comment by devarsh-mavani-19** [comment]
> Thanks a lot @devarsh-mavani-19 for finding this out! I had forgotten about this feature. Do we all agree on this?
> 
> * `Photographs taken on 2015-11-08` -> Hide
> * `Amavenita (ship, 2014)` -> Do not hide

yes

**Comment by nicolas-raoul** [comment]
@Discostu36 This bug still happens, I just tried on the latest Git code. Do you want to work on it? :-)

**Comment by Discostu36** [comment]
Sadly, I lack the skills. I have started learning Kotlin development, but I'm only at the very first baby steps (and don't have much time).

**Comment by sivaraam** [comment]
For those interested in picking this up, MR #4902 already has the changes to fix this. Only the tests need fixing IIRC. So, it could be used as a base.

I intend to get around to completing it. Do feel free to beat me to it, though 😉

**Comment by parneet-guraya** [comment]
I want to try ✋

**Comment by parneet-guraya** [comment]
Apart from filtering strings `needing` & `taken on` we have these filters 

- years that ends with XXX0s & are above 2000s are allowed i.e 2000, 2010, 2020.
- Otherwise we only allow `current` or `previous year`.

Now, this issue needs the second condition to be removed i.e any year should be allowed. But then it also doesn't make sense to have filter for decades (in the first point) right? 

### Solution: 

We can remove both of the filters and allow any year between 19XX & 20XX (inclusive) 

OR

Just have one filter to allow any year above >2000s instead of only `current` or `previous` year or in `decades` terms.

WDYT? @nicolas-raoul 

Also, api only works correctly under `prod` because the query (Amavenita) mentioned in the issue was returning empty result in `beta` , shouldn't be an issue right?

**Comment by nicolas-raoul** [comment]
@parneet-guraya If I understand correctly Sivaraam's comment above, only the unit tests need fixing. :-)

Further modifications might be worth a different issue.

Thanks!

**Comment by parneet-guraya** [comment]
> So, I'm starting to do you think if we could just take forward a combination of your changes (https://github.com/commons-app/apps-android-commons/pull/5761) and the one proposed in https://github.com/commons-app/apps-android-commons/pull/4902 in the next release and see how it goes 🤔

As per [comment](https://github.com/commons-app/apps-android-commons/issues/5749#issuecomment-2227279839), solution for this #4902 overlaps with the PR that was later merged #5761. I had to solve the conflict but solution made the other filters redundant that's why the comment (my comment above). 

>Now, this issue needs the second condition to be removed i.e any year should be allowed. But then it also doesn't make sense to have filter for decades (in the first point) right?

**Comment by sivaraam** [comment]
I get what you're trying to say @parneet-guraya. The fix done via #5761 does conflict with the PR #4902 significantly. I think the change does need a bit of tweak as a consequence. I've merged recent main and tweaked the fix accordingly. Hope this helps you show a way out of the conflict.

**Comment by parneet-guraya** [comment]
@sivaraam thanks for looping in and fixing the conflict. However, since the PR got conflicted with another one I found that even if I do solve it, changes made by get redundant. So, I rewrite the function again taking into consideration what you solved in the PR. Here it is #6167 if you think it can be improved further let me know :-)

**Comment by nicolas-raoul** [comment]
I uploaded many pictures using the fix branch, the issue described https://github.com/commons-app/apps-android-commons/issues/47 does not seem to be problematic anymore somehow.
